armory/art_experimental/attacks/kenansville_dft.py

```diff
@@ -0,0 +1,93 @@
+        In the paper, the attack assumed to have knowledge of word or phoneme locations
+        in the input. The attack implemented here assumes complete blackbox knowledge,
+        so the only options are: 1) modified the whole input or 2) modified subsequences
+        of the input with some probability.
+
+        param sample_rate: sample rate in Hz of inputs
+        param estimator: not used but necessary for interoperability with Armory/ART
+        param snr_db: the minimum SNR (in dB) to maintain
+        type snr_db: 'float'
+        param partial_attack: boolean to indicate if subsequences of the input are to be modified
+        param attack_len: length of subsequences to attack. Valid when partial_attack = True
+        param attack_prob: probability each subsequence will be attacked. Valid when partial_attack = True
+        param targeted: not used but necessary for interoperability with Armory
+        """
+        self.sample_rate = sample_rate
+        self.snr_db = snr_db
+        self.targeted = targeted
+        self.partial_attack = partial_attack
+        self.attack_len = attack_len
+        self.attack_prob = attack_prob
+
+    def _attack(self, x):
+        x_fft = np.fft.fft(x)
+        x_psd = np.abs(x_fft) ** 2
+        # sort by frequencies with increasing power
+        x_psd_ind = np.argsort(x_psd)
+        signal_db = 10 * np.log10(np.sum(x_psd))
+
+        """
+        The goal of the following search is to find all the
+        low power frequencies that can be discarded while
+        maintaining a minimum SNR.
+
+        If desired, the following coarse and fine search could
+        be replaced with binary search for faster convergence.
+        """
+        # coarse search
+        id = 2  # real signals have symmetry in positive and negative frequencies
+        noise = np.sum(x_psd[x_psd_ind[:id]])
+        noise_db = 10 * np.log10(noise)
+        while signal_db - noise_db > self.snr_db:
+            id *= 2
+            noise = np.sum(x_psd[x_psd_ind[:id]])
+            noise_db = 10 * np.log10(noise)
+
+        # fine search
+        id = int(id / 2)
+        noise = np.sum(x_psd[x_psd_ind[:id]])
+        noise_db = 10 * np.log10(noise)
+        while signal_db - noise_db > self.snr_db:
+            id += 2
+            noise = np.sum(x_psd[x_psd_ind[:id]])
+            noise_db = 10 * np.log10(noise)
+
+        # zero out low power frequencies
+        x_fft[x_psd_ind[: id - 2]] = 0
+        x_ifft = np.fft.ifft(x_fft)
+
+        return np.real(x_ifft).astype(np.float32)
+
+    def generate(self, x):
+        x_out = np.empty((len(x),), dtype=object)
+        for i, x_example in enumerate(x):
+            if self.partial_attack:
+                # split input into multiple segments and attack each with some probability
+                x_adv = np.zeros_like(x_example)
+                seg_len = self.attack_len
+                for j in range(len(x_example) // seg_len + 1):
```

Comment:
We should use a `ceil` method here.  Otherwise, this fails in the (rare) case where `len(x_example)` is an integer multiple of `seg_len`.

Reply:
Sounds good. Again, do you want to change it?